Core/Core/CourseSync/CourseSyncProgress/ViewModel/CourseSyncProgressViewModelItem.swift

```diff
@@ -0,0 +1,143 @@
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+//
+
+import SwiftUI
+
+extension CourseSyncProgressViewModel {
+
+    enum Cell: Hashable, Identifiable {
+        case empty(viewId: String)
+        case item(Item)
+
+        var id: String {
+            switch self {
+            case .empty(let viewId): return viewId
+            case .item(let item): return item.id
+            }
+        }
+    }
+
+    struct Item: Hashable, Identifiable {
+
+        /** The SwiftUI view ID. */
+        let id: String
+        let title: String
+        let subtitle: String?
+        var isCollapsed: Bool?
+        let cellStyle: ListCellView.ListCellStyle
+        let state: CourseSyncEntry.State
+
+        fileprivate(set) var collapseDidToggle: (() -> Void)?
+        fileprivate(set) var removeItemPressed: (() -> Void)?
+
+        static func == (lhs: CourseSyncProgressViewModel.Item, rhs: CourseSyncProgressViewModel.Item) -> Bool {
+            lhs.id == rhs.id &&
+            lhs.title == rhs.title &&
+            lhs.subtitle == rhs.subtitle &&
+            lhs.isCollapsed == rhs.isCollapsed &&
+            lhs.state == rhs.state
+        }
+
+        func hash(into hasher: inout Hasher) {
+            hasher.combine(id)
+            hasher.combine(title)
+            hasher.combine(subtitle)
+            hasher.combine(isCollapsed)
+            hasher.combine(state)
+        }
+    }
+}
+
+// MARK: - Mapping From Model Objects
+
+extension Array where Element == CourseSyncEntry {
+
+    func makeSyncProgressViewModelItems(interactor: CourseSyncProgressInteractor) -> [CourseSyncProgressViewModel.Cell] {
+        var cells: [CourseSyncProgressViewModel.Cell] = []
+
+        for (courseIndex, course) in enumerated() {
+            var courseItem = course.makeSyncProgressViewModelItem()
+            courseItem.collapseDidToggle = {
+                interactor.setCollapsed(selection: .course(courseIndex), isCollapsed: !(courseItem.isCollapsed ?? false))
+            }
+            cells.append(.item(courseItem))
+
+            if course.isCollapsed {
+                continue
+            }
+
+            if course.tabs.isEmpty {
+                cells.append(.empty(viewId: "course-\(course.id)-empty"))
+                continue
+            }
+
+            for (tabIndex, tab) in course.tabs.enumerated() {
+                var tabItem = tab.makeSyncProgressViewModelItem()
+                tabItem.collapseDidToggle = {
+                    interactor.setCollapsed(selection: .tab(courseIndex, tabIndex), isCollapsed: !(tabItem.isCollapsed ?? false))
+                }
+                cells.append(.item(tabItem))
+
+                guard tab.type == .files, !tab.isCollapsed else {
+                    continue
+                }
+
+                for file in course.files {
+                    var fileItem = file.makeSyncProgressViewModelItem()
```

Comment:
There a build warning here: `Variable 'fileItem' was never mutated; consider changing to 'let' constant`

Reply:
fixed.